### PR TITLE
Add support for statsmodels flavor

### PR DESCRIPTION
**Pull request by olbapjose**
## What changes are proposed in this pull request?

Add support for statsmodels flavor

## How is this patch tested?

Unit tests

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [X] Yes. Give a description of this change to be included in the release notes for MLflow users.

Model API now supports statsmodels, with autolog currently limited to function `fit`. Please be aware that some models do not natively support `predict`-ing and/or are not pickable (do not support pickle-based serialization), hence they cannot be saved.

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [X] `area/docs`: MLflow documentation pages
- [X] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [X] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: Local serving, model deployment tools, spark UDFs
- [ ] `area/server-infra`: MLflow server, JavaScript dev server
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, JavaScript, plotting
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [X] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Comment by olbapjose** [comment]
Excuse me, I wondered if someone could please have a look and give me some feedback. Sorry for the inconvenience and thanks a lot!

**Comment by pvillacorta-stratio** [comment]
Excuse me, just a gentle reminder, since I already updated this PR two weeks ago to resolve conflicts and make it mergeable again (sadly, more conflicts have arisen again after recent merges, I guess). 

If someone could please have a look to this PR, I would really appreciate it. I will be pleased to resolve conflicts again and incorporate all the changes that may be required. Thanks a lot!!

**Comment by tomasatdatabricks** [comment]
Hi @pvillacorta-stratio, I am taking a look now. Sorry for dropping the ball on this!

**Review by tomasatdatabricks — COMMENTED** [review]
This is a great PR, thanks for the contribution!

I left 2 minor comments, we can merge after those are addressed.

I've noticed the interface for time series models is a bit odd. I think it's ok to leave it as is for now, but it won't work with spark udfs in most cases. Spark udfs require that the the result has the same number of rows as the input.

File: mlflow/statsmodels.py [hunk 1]
```diff
@@ -0,0 +1,532 @@
...
+
+    def predict(self, dataframe):
+        from statsmodels.tsa.base.tsa_model import TimeSeriesModel
+
+        model = self.statsmodels_model.model
+        if isinstance(model, TimeSeriesModel):
+            # Assume the inference dataframe has columns "start" and "end", and just one row
+            # TODO: move this to a specific mlflow.statsmodels.tsa flavor? Time series models
+            # often expect slightly different arguments to make predictions
+            start_date = dataframe["start"][0]
```
[the remark this answers is not in the document]
Comment (pvillacorta-stratio) [reply mlflow/statsmodels.py:284]: Thanks, I will add a check to ensure the dataframe has 1 row and includes columns start and end.

File: mlflow/statsmodels.py [hunk 2]
```diff
@@ -0,0 +1,532 @@
...
+                        log_fn_args_as_params(original_method, args, kwargs)
+
+                    AutologHelpers.should_autolog = True
+
+                if auto_end_run:
+                    try_mlflow_log(mlflow.end_run)
+
+                return model
+
+            except Exception as e:
```
[the remark this answers is not in the document]
Comment (pvillacorta-stratio) [reply mlflow/statsmodels.py]: Done. Thanks a lot for your comments

[the remark this answers is not in the document]
Comment (pvillacorta-stratio) [reply mlflow/statsmodels.py]: Looks like the examples are failing due to pytorch examples, not related to this PR to statsmodels? Maybe I am missing something.

[the remark this answers is not in the document]
Comment (tomasatdatabricks) [reply mlflow/statsmodels.py]: I think there may be a flaky test, I'll rerun and merge it it passes.

File: mlflow/statsmodels.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (tomasatdatabricks) [reply mlflow/statsmodels.py:292]: Thanks!

**Review by tomasatdatabricks — APPROVED** [review]
lgtm! Thanks @pvillacorta-stratio!

File: mlflow/statsmodels.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (pvillacorta-stratio) [reply mlflow/statsmodels.py]: It seems the failing tests are
```
tests/examples/test_examples.py::test_mlflow_run_example[pytorch-params8] FAILED | Free disk space: 10.6 GiB
tests/examples/test_examples.py::test_mlflow_run_example[pytorch/BertNewsClassification-params18] FAILED | Free disk space: 4.1 GiB
```
I am wondering if this apparently unrelated PR is breaking something else that was not there in my previous attempts, but cannot find out right now.

[the remark this answers is not in the document]
Comment (harupy) [reply mlflow/statsmodels.py]: @pvillacorta-stratio 
Merging the current master should fix these issues.

[the remark this answers is not in the document]
Comment (pvillacorta-stratio) [reply mlflow/statsmodels.py]: Great, if you could please re-launch the PR as I don't have permissions to do it. Thanks a lot!

[the remark this answers is not in the document]
Comment (pvillacorta-stratio) [reply mlflow/statsmodels.py]: @tomasatdatabricks Finally fixed after merging the current master as @harupy said. Thanks guys.

**Pull request merged**